### PR TITLE
DrivAerML: Higher LR (7e-4) at 4L/512d

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The baseline uses lr=5e-4. On AirfRANS, lr=7e-4 with gradient clipping was the breakthrough LR. Higher LR at 4L/512d might find deeper basins faster within the 180-min budget.

## Instructions
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --agent shinobu \
  --wandb-name "shinobu/drivaerml-4L512d-lr7e4" \
  --wandb-group "drivaerml-4L512d-lr-sweep"
```

Report val_primary/surface_rel_l2_pct, best epoch, total epochs.

## Baseline
Current best: **4.619%** at epoch 256 (4L/512d, lr=5e-4, 267 epochs)
External target: <3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shinobu:

## Results

### Key Metrics

| Metric | Value | Baseline | Change |
|--------|-------|----------|--------|
| **val_primary/surface_rel_l2_pct** | **15.35%** (best, ep95) | 4.619% (ep256) | +232% (WORSE) |
| test_primary/surface_rel_l2_pct | 36.03% (final ep271) | — | — |
| Final val (ep271) | 35.70% | — | — |

**External target:** <3.71% (AB-UPT)

### Experiment Details

- **W&B run:** [l7ch474e](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/l7ch474e)
- **Epochs:** 271 (180-min timeout)
- **Peak memory:** ~6241 MiB
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw --lr 7e-4 --cosine-t-max 30 --no-use-ema \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --agent shinobu --wandb-name "shinobu/drivaerml-4L512d-lr7e4" \
  --wandb-group "drivaerml-4L512d-lr-sweep"
```

### What happened — NEGATIVE RESULT

**lr=7e-4 without gradient clipping is fundamentally unstable on DrivAerML.** The training exhibited repeated catastrophic gradient explosions that destroyed training progress, preventing the model from converging anywhere near the baseline.

#### Training trajectory (5 distinct phases):

1. **Phase 1 (ep1-22):** Normal convergence to 20.78% — promising early trajectory, faster than baseline
2. **Crash 1 (ep23):** Grad norms ~4.5, val spikes to 68.6%. Recovery took ~55 epochs to return to 20.78%
3. **Phase 2 (ep77-95):** Recovered and surpassed pre-crash best, reaching **15.35%** at ep95
4. **Crash 3 (ep97):** Catastrophic — grad norms exploded 12.6 → 296 → 4,572. Recovery took ~38 epochs back to 17%
5. **Terminal crash (ep145):** Grad norms 13K → 51K → 134K → 429K → 1M → 37.6M → 203M → 1.16B → 1.72B → **1.16T → 283T → 1.3 quadrillion**. Model never recovered. Oscillated 35-70% with permanently corrupted weights for remaining 126 epochs.

#### Key observations:
- **Pre-crash potential:** Before the terminal crash, the model was on a trajectory of ~1% improvement per cosine trough cycle. Had it stayed stable, it might have approached single-digit values by ep200+.
- **Crash severity escalation:** Each destabilization was progressively worse — grad norms went from ~4.5 (crash 1) to **trillions** (crash 5). The crashes are stochastic but inevitable at this LR without clipping.
- **Contrast with AirfRANS:** lr=7e-4 was the breakthrough LR on AirfRANS, but that success relied on gradient clipping (gc=0.5-1.5). Without clipping, the same LR is destructive on DrivAerML.
- **Contrast with baseline:** The baseline (lr=5e-4, same architecture) achieved 4.619% at ep256 with smooth, stable convergence over 267 epochs. No crashes.

### Suggested follow-ups

1. **lr=7e-4 + gradient clipping (gc=1.0 or gc=1.5)** on DrivAerML — this is the most promising direction. AirfRANS showed that gc + high LR is the winning combo. The pre-crash trajectory here (15.35% at ep95) suggests the LR has exploration power; it just needs clipping to survive cosine peaks.
2. **lr=6e-4** (interpolating between stable 5e-4 and unstable 7e-4) — might find a stable but more exploratory regime without needing clipping.
3. **lr=7e-4 + T_max=50** (longer cosine period) — the crashes correlate with cosine LR peaks at T_max=30. Longer periods mean the model spends less time at the dangerous peak LR, potentially improving stability.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2806 because ADVISOR: Closing — your result of 15.352% is far from the 4.619% baseline. lr=7e-4 caused a crash on DrivAerML, confirming this dataset needs conservative learning rates. Thanks for the data point, shinobu..
